### PR TITLE
[19.05] Run futurize for python2 cheetah templates, fix scope issue for py3 comprehensions

### PR DESCRIPTION
**Pull request by mvdbeek**
I think this should cover the python 2/3 errors I have seen so far

### CODE REVIEW

**Comment by mvdbeek** [comment]
Hah, turns out dict, set and generator comprehensions all have that problem on python 2, so I guess we don't need to fix that.

File: lib/galaxy/util/template.py [hunk 1]
```diff
@@ -15,4 +49,60 @@ def fill_template(template_text, context=None, **kwargs):
...
+                                     context=context,
+                                     retry=retry - 1,
+                                     compiler_class=compiler_class,
+                                     first_exception=first_exception
+                                     )
+        raise first_exception or e
+    except Exception as e:
+        if first_exception is None:
+            first_exception = e
+        if not futurized:
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py]:
```suggestion
        if not futurized and sys.version_info.major > 2:
```

Comment (mvdbeek) [reply lib/galaxy/util/template.py]: Thanks, I've added that to https://github.com/galaxyproject/galaxy/pull/7867/commits/8f692cbebd14bf4108ed050867aeb60bc54425d5

File: test/unit/tools/test_evaluation.py [hunk 2]
```diff
@@ -266,6 +266,7 @@ class MockTool(object):
 
     def __init__(self, app):
         self.profile = 16.01
+        self.python_template_version = 2.7
```
Comment (nsoranzo) [inline test/unit/tools/test_evaluation.py]:
```suggestion
        self.python_template_version = '2.7'
```

Comment (nsoranzo) [inline test/unit/tools/test_evaluation.py]:
```suggestion
        self.python_template_version = '2.7'
```

File: lib/galaxy/tools/__init__.py [hunk 3]
```diff
@@ -643,6 +644,14 @@ def parse(self, tool_source, guid=None, dynamic=False):
             message = template % (self.id, self.profile)
             raise Exception(message)
 
+        self.python_template_version = tool_source.parse_python_template_version()
+        if self.python_template_version is None:
+            # If python_template_version not specified we assume tools with profile versions >= 19.05 are python 3 ready
+            if self.profile >= 19.05:
+                self.python_template_version = 3.5
```
Comment (nsoranzo) [inline lib/galaxy/tools/__init__.py]:
```suggestion
                self.python_template_version = '3.5'
```

File: lib/galaxy/tools/__init__.py [hunk 4]
```diff
@@ -643,6 +644,14 @@ def parse(self, tool_source, guid=None, dynamic=False):
...
+            else:
+                self.python_template_version = 2.7
```
Comment (nsoranzo) [inline lib/galaxy/tools/__init__.py]:
```suggestion
                self.python_template_version = '2.7'
```

File: lib/galaxy/tools/actions/__init__.py [hunk 5]
```diff
@@ -925,7 +926,7 @@ def get_ext_or_implicit_ext(hda):
     return hda.ext
 
 
-def determine_output_format(output, parameter_context, input_datasets, input_dataset_collections, random_input_ext):
+def determine_output_format(output, parameter_context, input_datasets, input_dataset_collections, random_input_ext, python_template_version=3):
```
Comment (nsoranzo) [inline lib/galaxy/tools/actions/__init__.py]:
```suggestion
def determine_output_format(output, parameter_context, input_datasets, input_dataset_collections, random_input_ext, python_template_version='3'):
```

File: lib/galaxy/tools/parser/cwl.py [hunk 6]
```diff
@@ -150,6 +150,9 @@ def parse_requirements_and_containers(self):
     def parse_profile(self):
         return "16.04"
 
+    def parse_python_template_version(self):
+        return 3.5
```
Comment (nsoranzo) [inline lib/galaxy/tools/parser/cwl.py]:
```suggestion
        return '3.5'
```

File: lib/galaxy/tools/parser/xml.py [hunk 7]
```diff
@@ -483,6 +483,12 @@ def parse_profile(self):
         # - Enable buggy interpreter attribute.
         return self.root.get("profile", "16.01")
 
+    def parse_python_template_version(self):
+        python_template_version = self.root.get("python_template_version", None)
+        if python_template_version is not None:
+            python_template_version = float(python_template_version)
```
Comment (nsoranzo) [inline lib/galaxy/tools/parser/xml.py]:
```suggestion
            python_template_version = packaging.version.parse(python_template_version)
```

You'd need to also add `import packaging.version` at the top.

File: lib/galaxy/tools/parser/yaml.py [hunk 8]
```diff
@@ -162,6 +162,12 @@ def parse_tests_to_dict(self):
     def parse_profile(self):
         return self.root_dict.get("profile", "16.04")
 
+    def parse_python_template_version(self):
+        python_template_version = self.root_dict.get("python_template_version", None)
+        if python_template_version is not None:
+            python_template_version = float(python_template_version)
```
Comment (nsoranzo) [inline lib/galaxy/tools/parser/yaml.py]:
```suggestion
            python_template_version = packaging.version.parse(python_template_version)
```

You'd need to also add `import packaging.version` at the top.

File: lib/galaxy/util/template.py [hunk 9]
```diff
@@ -1,10 +1,45 @@
...
+    return CustomCompilerClass
+
+
+def fill_template(template_text,
+                  context=None,
+                  retry=10,
+                  compiler_class=Compiler,
+                  first_exception=None,
+                  futurized=False,
+                  python_template_version=3,
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py]:
```suggestion
                  python_template_version='3',
```

File: lib/galaxy/util/template.py [hunk 10]
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
...
+                lineno = last_stack.lineno - 1
+                module_code = t._CHEETAH_generatedModuleCode.splitlines()
+                module_code[lineno] = module_code[lineno].replace(replace_str, var_not_found)
+                module_code = "\n".join(module_code)
+                compiler_class = create_compiler_class(module_code)
+                return fill_template(template_text=template_text,
+                                     context=context,
+                                     retry=retry - 1,
+                                     compiler_class=compiler_class,
+                                     first_exception=first_exception
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py]: Add:
```python

                                     python_template_version=python_template_version,
```

File: lib/galaxy/util/template.py [hunk 11]
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
...
+            # through futurize and hope for the best.
+            module_code = t._CHEETAH_generatedModuleCode
+            module_code = futurize_preprocessor(module_code)
+            compiler_class = create_compiler_class(module_code)
+            return fill_template(template_text=template_text,
+                                 context=context,
+                                 retry=retry,
+                                 compiler_class=compiler_class,
+                                 first_exception=first_exception,
+                                 futurized=True
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py]: Add:
```python

                                     python_template_version=python_template_version,
```

File: lib/galaxy/util/template.py [hunk 12]
```diff
@@ -1,10 +1,45 @@
 """Entry point for the usage of Cheetah templating within Galaxy."""
+from __future__ import absolute_import
+
+import subprocess
+import sys
+import tempfile
+import traceback
+
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py:7]: Add:
```python
import packaging.version
```

File: lib/galaxy/util/template.py [hunk 13]
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
         raise TypeError("Template text specified as None to fill_template.")
     if not context:
         context = kwargs
-    return unicodify(Template(source=template_text, searchList=[context]))
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py:18]: Add:
```python
    if isinstance(python_template_version, str):
        python_template_version = packaging.version.parse(python_template_version)
```

File: lib/galaxy/util/template.py [hunk 14]
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
...
+    klass = Template.compile(source=template_text, compilerClass=compiler_class)
+    t = klass(searchList=[context])
+    try:
+        return unicodify(t)
+    except NotFound as e:
+        if first_exception is None:
+            first_exception = e
+        if python_template_version < 3 and retry > 0 and sys.version_info.major > 2:
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py]:
```suggestion
        if sys.version_info.major > 2 and python_template_version.release[0] < 3 and retry > 0:
```

File: lib/galaxy/util/template.py [hunk 15]
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
...
+                                     context=context,
+                                     retry=retry - 1,
+                                     compiler_class=compiler_class,
+                                     first_exception=first_exception
+                                     )
+        raise first_exception or e
+    except Exception as e:
+        if first_exception is None:
+            first_exception = e
+        if sys.version_info.major > 2 and python_template_version < 3 and not futurized:
```
Comment (nsoranzo) [inline lib/galaxy/util/template.py]:
```suggestion
        if sys.version_info.major > 2 and python_template_version.release[0] < 3 and not futurized:
```

File: test/unit/tools/test_fill_template.py [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    if sys.version_info.major > 2:
+        with pytest.raises(NotFound):
+            fill_template(LIST_COMPREHENSION_TEMPLATE, retry=0)
+    else:
+        template_str = fill_template(LIST_COMPREHENSION_TEMPLATE, retry=0)
+        assert template_str == 'echo 1\n'
+
+
+def test_fill_list_comprehension_template_2():
+    template_str = fill_template(LIST_COMPREHENSION_TEMPLATE, python_template_version=2, retry=1)
```
Comment (nsoranzo) [inline test/unit/tools/test_fill_template.py]:
```suggestion
    template_str = fill_template(LIST_COMPREHENSION_TEMPLATE, python_template_version='2', retry=1)
```

File: test/unit/tools/test_fill_template.py [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+        fill_template(SET_COMPR_TEMPLATE, retry=1)
+
+
+def test_gen_expr():
+    with pytest.raises(NotFound):
+        fill_template(GEN_EXPR_TEMPLATE, retry=1)
+
+
+def test_fix_template_two_to_three():
+    template_str = fill_template(TWO_TO_THREE_TEMPLATE, python_template_version=2, retry=1)
```
Comment (nsoranzo) [inline test/unit/tools/test_fill_template.py]:
```suggestion
    template_str = fill_template(TWO_TO_THREE_TEMPLATE, python_template_version='2', retry=1)
```

**Review by nsoranzo — COMMENTED** [review]
I'd prefer to parse `python_template_version` with `packaging.version.parse()` instead of casting with `float()`, see comments inline.

**Review by jmchilton — APPROVED** [review]

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by martenson** [comment]
merged forward

**Pull request merged**